src/nbl/video/IPhysicalDevice.cpp

```diff
@@ -90,4 +90,440 @@ bool IPhysicalDevice::validateLogicalDeviceCreation(const ILogicalDevice::SCreat
+    if (asset::isSignedFormat(a) != srcSignedFormat)
+        return false;
+    // Can't have less channels
+    if (asset::getFormatChannelCount(a) < srcChannels)
+        return false;
+
+    // Can't have less precision or value range
+    for (uint32_t c = 0; c < srcChannels; c++)
+    {
+        double mina = asset::getFormatMinValue<double>(a, c),
+            minb = asset::getFormatMinValue<double>(b, c),
+            maxa = asset::getFormatMaxValue<double>(a, c),
+            maxb = asset::getFormatMaxValue<double>(b, c);
+
+        // return false if a has less precision (higher precision delta) than b
+        // check at 0, since precision is non-increasing
+        // also check at min & max, since there's potential for cross-over with constant formats
+        if (getFormatPrecisionAt(a, c, 0.0) > getFormatPrecisionAt(b, c, 0.0)
+                || getFormatPrecisionAt(a, c, srcMin[c]) > getFormatPrecisionAt(b, c, srcMin[c])
+                || getFormatPrecisionAt(a, c, srcMax[c]) > getFormatPrecisionAt(b, c, srcMax[c]))
+            return false;
+        // return false if a has less range than b
+        if (mina > minb || maxa < maxb)
+            return false;
+    }
+    return true;
+}
+
+// Returns true if 'a' is a better fit than 'b' (for tie breaking)
+// Tie-breaking rules:
+// - RGBA vs BGRA matches srcFormat
+// - Maximum precision delta is smaller
+// - Value range is larger
+bool isFormatBetterFit(asset::E_FORMAT a, asset::E_FORMAT b, bool srcBgra, uint32_t srcChannels, double srcMin[], double srcMax[])
+{
+    bool curBgraMatch = asset::isBGRALayoutFormat(a) == srcBgra;
+    bool prevBgraMatch = asset::isBGRALayoutFormat(b) == srcBgra;
+
+    // if one of the two fits the original bgra better, use that
+    if (curBgraMatch != prevBgraMatch)
+        return curBgraMatch;
+
+    // Check precision deltas
+    for (uint32_t c = 0; c < srcChannels; c++)
+    {
+        // View comments above about value selection
+        // Pick the max precision delta for each format
+        double deltaA = std::max(std::max(getFormatPrecisionAt(a, c, 0.0), getFormatPrecisionAt(a, c, srcMin[c])), getFormatPrecisionAt(a, c, srcMax[c]));
+        double deltaB = std::max(std::max(getFormatPrecisionAt(b, c, 0.0), getFormatPrecisionAt(b, c, srcMin[c])), getFormatPrecisionAt(b, c, srcMax[c]));
+
+        // if one of the two has a better max precision delta, use that
+        if (deltaA != deltaB)
+            return deltaA < deltaB;
+    }
+
+    // Check value range
+    for (uint32_t c = 0; c < srcChannels; c++)
+    {
+        double mina = asset::getFormatMinValue<double>(a, c),
+            minb = asset::getFormatMinValue<double>(b, c),
+            maxa = asset::getFormatMaxValue<double>(a, c),
+            maxb = asset::getFormatMaxValue<double>(b, c);
+
+        // return true if a has better range than b
+        if (mina < minb && maxa > maxb)
+            return true;
+    }
```

Comment:
ok, what we want to avoid is using a format which has less "quantifiable values" in the source format's range,

so we should probably take the one that has the smallest "waste" given source input/output ranges
```
double waste = 0.0;
for (auto c=0; c<srcChannelCount; c++)
{
   // you can assert that the ranges of the dest are greater
   waste += (srcMin[c]-dstMin[c])/dstPrecision(srcMin[c],c);
   waste += (dstMax[c]-srcMax[c])/dstPrecision(srcMax[c],c);
}
```

Then tie break using the format with the lowest "waste"